syntheseus/search/algorithms/best_first/base.py

```diff
@@ -53,72 +120,85 @@ def _run_from_graph_after_setup(self, graph: GraphType) -> int:
-                    tie_breaker += 1
-                else:
-                    action = f"discarded {priority_change_str}"
+            # Take nodes from the priority queue until a node eligible for expansion is found
+            num_popped = 0
+            while len(queue) > 0:
+                pq_item = queue.pop_item()
+                num_popped += 1
+
+                # Do a few checks
+                assert pq_item.priority < math.inf, "inf priority should not be in the queue"
+                assert math.isclose(
+                    pq_item.priority, self.priority_function(pq_item.item, graph)
+                ), "priority is not up-to-date. This should not happen."
+
+                if self.node_eligible_for_queue(pq_item.item, graph):
+                    node = pq_item.item
+                    break
             else:
-                # Visit node
-                new_nodes = list(self.visit_node(node, graph))
-
-                # Update node values
-                nodes_updated = self.set_node_values(new_nodes + [node], graph)
-
-                # Add new eligible nodes to the queue, since
-                # their priority may have changed.
-                # dict.fromkeys is to preserve order and uniqueness
-                for updated_node in dict.fromkeys(new_nodes + list(nodes_updated)):
-                    if self.node_eligible_for_queue(updated_node, graph):
-                        updated_node_priority = self.priority_function(updated_node, graph)
-                        if updated_node_priority < math.inf:
-                            heapq.heappush(
-                                queue, (updated_node_priority, tie_breaker, updated_node)
-                            )
-                            tie_breaker += 1
-
-                # Log str
-                action = f"visited, {len(new_nodes)} new nodes created, {len(nodes_updated)} nodes updated)"
+                logger.log(log_level, "No eligible node found. Stopping search.")
+                break
+
+            # Visit node
+            new_nodes = list(self.visit_node(node, graph))
+
+            # Update node values
+            nodes_updated = self.set_node_values(new_nodes + [node], graph)
+
+            # Add new nodes and updated nodes to the queue, since their priority may have changed.
+            # dict.fromkeys is to preserve order and uniqueness
+            for updated_node in dict.fromkeys(new_nodes + list(nodes_updated)):
+                if self.node_eligible_for_queue(updated_node, graph):
+                    updated_node_priority = self.priority_function(updated_node, graph)
+                    already_in_queue_at_correct_priority = (
+                        updated_node in queue._entry_finder
+                        and math.isclose(
+                            queue._entry_finder[updated_node].priority, updated_node_priority
+                        )
+                    )
```

Comment:
Maybe `PriorityQueue` could have `__contains__` and `get_priority` helpers implemented, so that the outside code doesn't need to access the private `_entry_finder` attribute?

Reply:
Good suggestions: done!